feat: allow update function to trigger file deletion

stefanbuck opened the pull request:
Fixes #129 

I raised this PR to discuss some challenges I phased with TypeScript see comments.

stefanbuck commented on src/create-tree.ts:
```diff
@@ -71,7 +95,7 @@ export async function createTree(
-      if (result === null || typeof result === "undefined") continue;
+      if (result === null || typeof result === "undefined" || typeof result === "symbol") continue;
       tree.push(
         await valueToTreeObject(octokit, ownerOrFork, repo, path, result)
```
TS says:
```
Argument of type 'string | File | Symbol' is not assignable to parameter of type 'string | File'.
Type 'Symbol' is not assignable to type 'string | File'
```

BUT, the condition above prevents this code path if result is of type symbol or am I missing something here?

gr2m replied:
> the condition above prevents this code path if result is of type symbol or am I missing something here

if that happenes and TS is not smart enough to get it, I'd suggest to add `// @ts-expect-error - explanation here` to make TS calm down about it

stefanbuck commented on src/types.ts:
```diff
@@ -54,7 +54,7 @@ export type UpdateFunctionFile =
 export type UpdateFunction = (
   file: UpdateFunctionFile
-) => string | File | null | Promise<string | File | null>;
+) => string | File | null | Symbol | Promise<string | File | null | Symbol>;
```
If the update function returns `null` it means no file changes.
If the update function return a `symbol` it triggers a file deletion.

stefanbuck replied:
Is there a way to type the Symbol to ensure it is the actual `deleteFile` symbol?

gr2m replied:
I don't know, I actually never worked with types for Symbols

stefanbuck commented on src/index.ts:
```diff
@@ -2,6 +2,7 @@ import type { Octokit } from "@octokit/core";
 
 import { composeCreatePullRequest } from "./compose-create-pull-request";
 import { VERSION } from "./version";
+export { deleteFile } from "./constants";
```
### Usage

```js
import {  deleteFile } from 'octokit-plugin-create-pull-request';

...

files: {
  "file/to/delete.txt": ({ exists, encoding, content }) => {
    const fileContent = Buffer.from(content, encoding).toString("utf-8")

    if (fileContent.includes('abc')) {
      // trigger file deletion
      return deleteFile
    }

    // do not alter file content
    return null;
  }
}
```

gr2m replied:
Nice, I like it 👍🏼

stefanbuck commented on src/create-tree.ts:
```diff
@@ -20,6 +21,8 @@ export async function createTree(
     const value = changes.files[path];
 
+    // TODO use Symbol here which will result in a breaking change
+    // if (value === deleteFile) {
```
or we keep it as is at the drawback of having inconsistency between return values from an update function vs this code snipped

```
files: {
  "file/to/delete.txt": null
}

```

gr2m replied:
I'd rather go with a breaking change

stefanbuck replied:
Done. Documented in the readme and commit message

stefanbuck commented on src/create-tree.ts:
```diff
@@ -62,6 +63,28 @@ export async function createTree(
+            });
+            continue;
+          } catch (error) {
+            // istanbul ignore next
```
I thought this is fine since I saw basically the same for other catch blocks, but please let me know if you would prefer a test case for this.

stefanbuck commented:
@gr2m this PR is ready for review. 

- I addressed your comments regarding TS 
- Added a unit test
- Updated the readme
- Added a breaking change note in the commit footer

gr2m commented on README.md:
````diff
@@ -38,7 +38,10 @@ Install with `npm install @octokit/core octokit-plugin-create-pull-request`. Opt
 ```js
 const { Octokit } = require("@octokit/core");
-const { createPullRequest } = require("octokit-plugin-create-pull-request");
+const {
+  createPullRequest,
+  deleteFile,
+} = require("octokit-plugin-create-pull-request");
````
Seeing these two imports, I really don't like `deleteFile`. At first sight it suggests that it's a separate method to delete a file, don't you think?

Node.js has a convention to uses a `k` prefix for constants, see example: https://github.com/nodejs/node/blob/4bb06dbd0a87e5b2430780ddceec08015f8c496a/lib/_http_client.js#LL107C7-L107C13

Maybe we could use `kDeleteFile`? Or how about `DELETE_FILE`?

stefanbuck replied:
Good shout. I changed it to `DELETE_FILE` which is well understood in our industry

gr2m reviewed (commented):
Besides the note about the `deleteFile` export name, I really love your pull request, thank you so much!! I'm okay with merging it as is, but wanted to at least try to come up with a better name than `deleteFile`. Let me know what you think

gr2m reviewed (approved):
Thank you!
